### PR TITLE
Custom actions

### PR DESCRIPTION
**Pull request by mlucool**
Fixes: https://github.com/jupyterlab/jupyterlab-git/issues/584

This adds the ability for users to set a custom class to take actions. As of now, we only have a post_init action, but this can be trivially extended to give more power to end users who have non-generic needs.

Example `~/.jupyter/jupyter_notebook_config.json`:
```
  "JupyterLabGitConfig": {
      "custom_git_actions_class": "org.dir.MyClass"
  }
```

### CODE REVIEW

**Comment by mlucool** [comment]
I'm add to add documentation to the README and test cases if this is the right design. Thought I'd bring this up for feedback first

**Comment by fcollonval** [comment]
@mlucool Thanks for this PR. Using server configuration is safe enough - managed installation will keep control :wink: 

Regarding the implementation, I suggest to not be as generic for two reasons:
- We don't want to deal with complex custom Python code plugged-in
- We want to keep some control over those side-effects - like be able to implement a timeout feature around them.

So I propose to use the following configuration to a list of subprocess commands to be executed in the repository folder:

```js
{
  "JupyterLabGit": {
    "actions": {
      "postInit": [{
        "command": Array<string>, // subprocess command
        "args": Dict[string, string]  // subprocess args
      }]
    }
  }
}
```

Some comments on the schema:

- Use `JupyterLabGit` shorter name - we know that we are speaking about configuration
- Add a subentry `actions` so this will allow to organize the configuration parameters - if it grows in the future
- Add actions subentry with clear name about when it is gonna be executed (e.g. you could add a `postClone` entry)

What do you think?

> To make it works easily with `Configurable`, I suggest to use `AutoInstance` implemented in [jupyter-project](https://github.com/fcollonval/jupyter-project/blob/master/jupyter_project/config.py).

**Comment by mlucool** [comment]
> Regarding the implementation, I suggest to not be as generic for two reasons:

I don't totally agree, but I'm happy to oblige. One key reason is that the API is very hard to establish - but maybe we just promise to give the script nothing and we call it in the right cwd. I also find in process error handling mostly easier. Still, we ca go with arbitary shell commands only, but I think we just support anything a user puts and use `shlex` to break it up or use the `shell=True` ( i.e. '/my/script.py --foo' is a valid command to pass).

> Some comments on the schema:

These all seem reasonable

**Comment by fcollonval** [comment]
> but I think we just support anything a user puts and use shlex to break it up or use the shell=True ( i.e. '/my/script.py --foo' is a valid command to pass).

Good point, this will ease the setting.

**Comment by mlucool** [comment]
@fcollonval updated. Any more feedback or I will finish cleaning this up (more doc, seeing if its easy enough to test). I can add this to places besides init (e.g. clone) as well.

**Comment by mlucool** [comment]
@fcollonval did you have further feedback?

File: jupyterlab_git/__init__.py [hunk 1]
```diff
@@ -16,6 +34,8 @@ def _jupyter_server_extension_paths():
 def load_jupyter_server_extension(nbapp):
     """Load the Jupyter server extension.
     """
-    git = Git(nbapp.web_app.settings['contents_manager'])
+
+    user_custom_actions = JupyterLabGit(config=nbapp.config).actions
+    git = Git(nbapp.web_app.settings['contents_manager'], user_custom_actions)
```
Comment (fcollonval) [inline jupyterlab_git/__init__.py:39]: I would pass the all configuration object to anticipate other configuration settings.

```suggestion
    config = JupyterLabGit(config=nbapp.config)
    git = Git(nbapp.web_app.settings['contents_manager'], config)
```

File: jupyterlab_git/git.py [hunk 2]
```diff
@@ -870,14 +872,33 @@ async def init(self, current_path):
...
+            actions_list = self.user_custom_actions[name]
+            for action in actions_list:
+                # We trust the actions as they were passed via a config and not the UI
+                code, stdout, stderr = await execute(
+                    shlex.split(action), cwd=cwd
+                )
+                # After any failure, stop
+                if code != 0:
+                    break
+        return code, stdout, stderr
```
Comment (fcollonval) [inline jupyterlab_git/git.py:900]: Opened question: Should the stdout be concatenate for all actions?

File: jupyterlab_git/git.py [hunk 3]
```diff
@@ -870,14 +872,33 @@ async def init(self, current_path):
         Execute git init command & return the result.
         """
         cmd = ["git", "init"]
+        cwd = os.path.join(self.root_dir, current_path)
         code, _, error = await execute(
-            cmd, cwd=os.path.join(self.root_dir, current_path)
+            cmd, cwd=cwd
         )
+        if code == 0:
+            code, _, error = await self._maybe_run_actions('post_init', cwd)
```
Comment (fcollonval) [inline jupyterlab_git/git.py:880]: Opened question: Should stdout logged to help debugging?

File: jupyterlab_git/git.py [hunk 4]
```diff
@@ -870,14 +872,33 @@ async def init(self, current_path):
...
+        code = 0
+        stdout = None
+        stderr = None
+        if name in self.user_custom_actions:
+            actions_list = self.user_custom_actions[name]
+            for action in actions_list:
+                # We trust the actions as they were passed via a config and not the UI
+                code, stdout, stderr = await execute(
+                    shlex.split(action), cwd=cwd
+                )
```
Comment (fcollonval) [inline jupyterlab_git/git.py:896]:
```suggestion
                try:
                    code, stdout, stderr = await execute(
                        shlex.split(action), cwd=cwd
                    )
                except CalledProcessError:
                    break
```

Should we also protect against subprocess failure or do we consider it corrupt the base git action?

File: jupyterlab_git/__init__.py (hunk 1 above)
Comment (mlucool) [reply jupyterlab_git/__init__.py]: Addressed. Not I had to name it `_config` in `git.py` or a test was failing.

File: jupyterlab_git/git.py (hunk 3 above)
Comment (mlucool) [reply jupyterlab_git/git.py]: Do we have any examples of logging? For now, I sent all data (stdout/stderr/code/command) back to the front end where we can later show whatever we'd like to the user.

File: jupyterlab_git/git.py (hunk 4 above)
Comment (mlucool) [reply jupyterlab_git/git.py]: I caught all exceptions here and bailed.

**Comment by mlucool** [comment]
@fcollonval this is ready for re-review.

> For the test, could you test also a failure of execute with a non-zero code and with the emission of an exception? This is quite critical and will help use ensuring robustness of the extension.

Tested no post_init, post_init that works, and post_init that fails

File: jupyterlab_git/git.py [hunk 5]
```diff
@@ -872,13 +874,50 @@ async def init(self, current_path):
...
+                    actions.append({
+                        'cmd': action,
+                        'code': code,
+                        'stdout': stdout,
+                        'stderr': stderr
+                    })
+                    # After any failure, stop
+                except Exception as e:
+                    code = 1
+                    action.append({
```
Comment (fcollonval) [inline jupyterlab_git/git.py:911]:
```suggestion
                    actions.append({
```

This is obviously not tested 😜

File: jupyterlab_git/tests/test_init.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        } == actual_response
+
+
+@pytest.mark.asyncio
+async def test_init_and_post_init_fail():
+    with patch("jupyterlab_git.git.execute") as mock_execute:
+        # Given
+        mock_execute.side_effect = [
+            maybe_future((0, "", "")),
+            maybe_future((1, "", "not_there: command not found")),
```
Comment (fcollonval) [inline jupyterlab_git/tests/test_init.py:63]: You can test the `Exception` emission by duplicating this test and set an Expection object as second element in the list.

File: jupyterlab_git/git.py (hunk 5 above)
Comment (mlucool) [reply jupyterlab_git/git.py]: Good catch!

File: jupyterlab_git/tests/test_init.py (hunk 6 above)
Comment (mlucool) [reply jupyterlab_git/tests/test_init.py:63]: That tip was a big time saver - tested!

**Comment by mlucool** [comment]
@fcollonval I believe all review feedback is taken care of

**Comment by fcollonval** [comment]
@meeseeksdev backport to 0.11.x

**Pull request merged**